### ISSUE TITLE
secret cannot be deleted even if the sa was deleted

### ISSUE THREAD

**Issue by gyliu513** [istio/istio#9349]
When I enabled Citadel, after I create a sa, then a secret named as `istio.${sa}` will be created.

But after the sa was deleted, the secret was still there and after I delete this secret manually, it was created again.

I think Citadel should delete the related secret when the sa was deleted.

```
root@gyliu-icp-8:~/cases/istio-1.0.2/install/kubernetes# kubectl create sa aaa
serviceaccount/aaa created
root@gyliu-icp-8:~/cases/istio-1.0.2/install/kubernetes# kubectl get secrets
NAME                  TYPE                                  DATA      AGE
aaa-token-j89t7       kubernetes.io/service-account-token   3         2s
default-token-gjc8d   kubernetes.io/service-account-token   3         41d
istio.aaa             istio.io/key-and-cert                 3         1s
istio.default         istio.io/key-and-cert                 3         42m
root@gyliu-icp-8:~/cases/istio-1.0.2/install/kubernetes# kubectl delete sa aaa
serviceaccount "aaa" deleted
root@gyliu-icp-8:~/cases/istio-1.0.2/install/kubernetes# kubectl get secrets
NAME                  TYPE                                  DATA      AGE
default-token-gjc8d   kubernetes.io/service-account-token   3         41d
istio.aaa             istio.io/key-and-cert                 3         1s
istio.default         istio.io/key-and-cert                 3         42m
root@gyliu-icp-8:~/cases/istio-1.0.2/install/kubernetes# kubectl delete secrets istio.aaa
secret "istio.aaa" deleted
root@gyliu-icp-8:~/cases/istio-1.0.2/install/kubernetes# kubectl get secrets
NAME                  TYPE                                  DATA      AGE
default-token-gjc8d   kubernetes.io/service-account-token   3         41d
istio.aaa             istio.io/key-and-cert                 3         2s
istio.default         istio.io/key-and-cert                 3         42m
```

/cc @elevran @myidpt

**Comment by myidpt** [comment]
Thanks for catching this.
Citadel recreated the secret because it still found the service account. My guess is that the apiserver has some delay that the deleted service account is not immediately removed.
https://github.com/istio/istio/blob/release-1.0/security/pkg/pki/ca/controller/secret.go#L289
I will try to add some delay and see if it fixes the bug.

**Comment by myidpt** [comment]
Update: The function to retrieve service account does not return nil service account when the service account is removed. Existing code relies on this  causes recreation of secret. We should rely on the returned error code instead to check if the service account exists.

**Comment by myidpt** [comment]
If I recall correctly, it’s fixed on all branches other than release-1.0.

**Comment by gyliu513** [comment]
@myidpt thanks, can we merge it to release-1.0, so that the fix is available when we release 1.0.6 if we have that version? ;-)

**Comment by myidpt** [comment]
Sure. I will do that.
